File: CPP/OCP_OTHER_CODING_PROBS/DupStrings.cpp
// Find duplicated strings (words) in an array of words. Print the words that recur, as well as the total
// number of times a word recurs at least once.

#include <iostream>
#include <array>
#include <list>
#include <map>
#include <string>

using namespace std;

class DupStrings {

    public:

        static void dupInfo(list<string> wl) {
            
            map<string, int> wordOccurs;

            for (const string& thisWord: wl)
                    wordOccurs[thisWord]++;

            int repeats = 0;
            for (map<string, int>::iterator iter = wordOccurs.begin(); iter != wordOccurs.end();
                                                                       iter++ ) {

                string thisWord = iter->first;
                int count = iter->second;

                if (count > 1) {

                    printf("%s : %d\n", thisWord.c_str(), count);

                    repeats += 1;
                }
            }

            printf( "Total words with 2+ occurences: %d\n", repeats );

        }

};

//////////////////////////////////////////////////////////////////////////////////////

int main(int argc, char ** args) {

        list<string> testWords{ "delhi", "bombay", "delhi", "dodballapur", "bombay",
                                  "calcutta", "bombay"};

        DupStrings::dupInfo(testWords);

}

File: RUST/OCP_OTHER_CODING_PROBS/DupStrings.rs
// Find duplicated Strings (words) in an array of words. Print the words that recur,
// as well as the total number of times a word recurs at least once.

use std::collections::HashMap;

fn dup_info(wl: Vec<&str>) {
            
    let mut word_occurs = HashMap::<&str, i32>::new();

    for this_word in wl.iter() {

        let mut count = 0;
        if word_occurs.contains_key(this_word) {
            count = word_occurs[this_word];
        }

        count += 1;
        word_occurs.insert(this_word, count);
    }

    let mut repeats = 0;

    for (this_word, count) in &word_occurs {

        if count > &1 {

            println!("{} : {}\n", this_word, count);

            repeats += 1;
        }
    }

    println!( "Total words with 2+ occurences: {}\n", repeats );

}


//////////////////////////////////////////////////////////////////////////////////////

fn main() {

    let test_words = vec![ "delhi", "bombay", "delhi", "dodballapur", "bombay", "calcutta", "bombay"];

    dup_info(test_words);
}
